feature/20692-treemap-small-nodes-grouping

hubertkozik opened the pull request:
Added new treemap feature, the grouping of small leaf nodes, configured by the [treemap.cluster](https://api.highcharts.com/highcharts/series.treemap.cluster) option.

---

Closes #20692.

- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206668318846607

KacperMadej reviewed (changes requested):
- [x] Click on the grouped after drilling down to the top left category in highcharts/plotoptions/treemap-grouping-advanced is causing an error in the console.
- [ ] Both bottom right groups only partially fill the available space (like 50%) after a drilldown.
![image](https://github.com/user-attachments/assets/1aafc90c-fd21-467c-b35e-607cdc6ac0e5)
![image](https://github.com/user-attachments/assets/b2ef6831-fb93-42b2-9be5-38a8d7e174b7)
- [x] drawing `dataLabels` in the advanced demo takes over 8 seconds
![image](https://github.com/user-attachments/assets/0fb86560-2033-496c-8dd8-b6cf243a2787)

hubertkozik commented on ts/Series/Treemap/TreemapSeries.ts:
```diff
@@ -427,6 +427,158 @@ class TreemapSeries extends ScatterSeries {
+                    area = a * b;
+
+                if (
+                    (
+                        compareHeight &&
+                        (
+                            a < pixelWidth ||
+                            b < pixelHeight ||
+                            area < thresholdArea
+                        )
+                    ) ||
+                    (
+                        !compareHeight &&
+                        (
+                            a < pixelWidth ||
+                            b < pixelWidth ||
+                            area < thresholdArea
+                        )
+                    )
```
Right, I didn't simplify it! Fixed.

bre1470 commented on ts/Series/Treemap/TreemapNode.ts:
```diff
@@ -37,6 +37,7 @@ class TreemapNode {
     childrenTotal = 0;
     ignore?: boolean;
     isLeaf?: boolean;
+    isGroup: boolean = false;
```
You can make it optional and skip the initial assignment.

bre1470 commented on ts/Series/Treemap/TreemapSeriesDefaults.ts:
```diff
@@ -209,8 +209,25 @@ const TreemapSeriesDefaults: TreemapSeriesOptions = {
+         * `tooltip.formatter` callback function and `point.isGroupNode` flag.
+         *
+         * @type      {string}
+         * @default   + {point.groupedPointsAmount} more...
```
C&P problem with `+` and `more...`. This should probably be a string.

bre1470 commented on ts/Series/Treemap/TreemapSeriesOptions.d.ts:
```diff
@@ -106,6 +106,55 @@ export interface TreemapSeriesLevelColorVariationOptions {
+    /**
+     * The pixel threshold width of area, which is used in Treemap grouping.
+     *
+     * @type {number}
+     *
+     * @since next
+     *
+     * @product highcharts
+     *
```
Unnecessary spacing as in others as well. Keep it compact for one-liners.

bre1470 commented on ts/Series/Treemap/TreemapSeriesOptions.d.ts:
```diff
@@ -591,7 +660,20 @@ export interface TreemapSeriesOptions extends ScatterSeriesOptions {
+     * `tooltip.formatter` callback function and `point.isGroupNode` flag.
+     *
+     * @type      {string}
+     * @default   + {point.groupedPointsAmount} more...
```
The plus-more-thing ones more.

bre1470 commented on ts/Series/Treemap/TreemapSeriesDefaults.ts:
```diff
@@ -542,7 +559,69 @@ const TreemapSeriesDefaults: TreemapSeriesOptions = {
+         * @since next
+         * @product highcharts
+         * @apioption plotOptions.treemap.groupAreaThreshold.pixelHeight
+         *
+         */
+
+        /**
+         * The name of the point of grouped nodes shown in the tooltip,
+         * dataLabels, etc.
+         *
+         * @type {string}
+         *
+         * @since next
+         *
```
Unnecessary spacing

KacperMadej reviewed (approved):
Looks good to me.
Isn't the format "+ X more..." a bit misleading? I would expect to the group to contain **over** X and not exactly X.

MarkusBarstad reviewed (approved):
Very well done! 🚀

bre1470 reviewed (approved):
In the grouping demo the rest tooltip says "New tooltip content", which is a bit strange.
And I agree with what Kacper wrote.

Otherwise it looks good. 👍

TorsteinHonsi commented on samples/highcharts/plotoptions/treemap-grouping-advanced/demo.js:
```diff
@@ -0,0 +1,118 @@
+        'https://cdn.jsdelivr.net/gh/highcharts/highcharts@8a1f8f0eb4/samples/data/africa-export-2021.json'
+    ).then(response => response.json());
+
+    const lightenDarkenColor = (col, amt) =>
```
We have the [Highcharts.Color.brighten](https://api.highcharts.com/class-reference/Highcharts.Color#brighten) function that we use in some demos that looks a bit friendlier...

TorsteinHonsi reviewed (changes requested):
Thanks, great work! I added some comments inline. 

- [x] Also, I think this feature deserves a mention in the `docs/chart-and-series-types/treemap.md` article.

- [x] I addition, did you consider adding a feature to click the grouped item to expand it? I am not sure how that would work, if it should expand to the parent level or only the grouped items. But this would be equivalent to clicking a [marker cluster in scatter charts](https://jsfiddle.net/gh/get/library/pure/highcharts/highcharts/tree/master/samples/highcharts/marker-clusters/basic/), where the cluster dissolves into the individual points.

hubertkozik commented on ts/Series/Treemap/TreemapSeriesDefaults.ts:
```diff
@@ -542,7 +559,67 @@ const TreemapSeriesDefaults: TreemapSeriesOptions = {
+         * @since next
+         * @product highcharts
+         */
+        name: 'Grouped nodes'
```
Agree, removed default name and added info in API description.

TorsteinHonsi reviewed (commented):
Thanks so far! I've been fiddling a bit with the traversing and have some comments/questions:

1. Consider highcharts/plotoptions/treemap-grouping-advanced. If you click the "All machinery" group, the nested nodes expand, but they seem to do so without animation. As far as I can see, this is contrary to the `treemap-large` demo, where animation occurs. 
2. Next, click the +168 tile. Also no animation.
3. The breadcrumbs now say +112. But I clicked +168...
4. Click Machinery again in the breadcrumbs. Now the +168 tile takes up som 30-40% of the plot area, unlike in step 1 where it was much smaller. 
5. Click +168 again. Now the breadcrumbs say +153.

All in all we may have to work a bit on animation. For comparison, see the highcharts/marker-clusters/basic demo, where all animations flow nicely, all points originating in the cluster. But this may not be unique to this PR or the new feature, but maybe a treemap issue by itself.

hubertkozik commented:
@TorsteinHonsi I fixed the animation problem, turns out it was the `animationLimit` for this series in this demo.

Also, I fixed the problem of grouped node wrong sizing - it was much more complicated and I figured out, that the previous implementation of grouping was done in the wrong place, I refactored it one more time.

Sadly, now after traversing into grouped nodes their children sometimes look strange, but it's a correct layout behaviour - not sure if we should improve that. For example, after traversing into a grouped node in "Machinery" group there are some empty points, due to layout change/re-calculation, but only good points are shown on the chart.

TorsteinHonsi commented:
> To be honest I am not sure if we should provide the possibility to traverse into grouped points, due to algorithm calculations it gets very complicated.

@hubertkozik Could you make a team decision on this in the core group? The client probably has a preference. If we can't decide clearly for one solution or the other, we could introduce an option.

bre1470 commented on ts/Series/Treemap/TreemapSeriesDefaults.ts:
```diff
@@ -542,7 +561,91 @@ const TreemapSeriesDefaults: TreemapSeriesOptions = {
+     * @since next
+     *
+     * @product highcharts
+     * @apioption plotOptions.treemap.groupAreaThreshold
```
You could skip `@apioption` as it is part of an `@optionparent`-tree.

bre1470 reviewed (approved):
Great job! Looks good. 👍

TorsteinHonsi commented on samples/highcharts/plotoptions/treemap-grouping-advanced/demo.js:
```diff
@@ -0,0 +1,117 @@
+(async () => {
+
+    const dataJson = await fetch(
```
Just `data`, it is no longer JSON after the JSON string has been parsed.
```suggestion
    const dataJson = await fetch(
```

TorsteinHonsi commented on ts/Series/Treemap/TreemapSeriesDefaults.ts:
```diff
@@ -210,7 +212,21 @@ const TreemapSeriesDefaults: TreemapSeriesOptions = {
+         * @default   '+ {point.groupedPointsAmount} more...'
+         * @apioption tooltip.clusterFormat
+         */
+        clusterFormat: '<span>+ {point.groupedPointsAmount} more...' +
```
Why the span?

hubertkozik replied:
Right, span not needed.

TorsteinHonsi commented on ts/Series/Treemap/TreemapSeriesDefaults.ts:
```diff
@@ -542,7 +558,120 @@ const TreemapSeriesDefaults: TreemapSeriesOptions = {
+     * nodes below a certain square area threshold in pixels. If the square area
+     * of a point becomes smaller than the specified threshold, determined by
+     * the `pixelWidth` and/or `pixelHeight` options, then this point is moved
+     * into group point per series.
```
Did you mean into _one_ group point?
```suggestion
     * into one group point per series.
```

TorsteinHonsi reviewed (changes requested):
Great work! Just some minor details/questions.
